solutionbox/code_free_ml/mltoolbox/code_free_ml/data/trainer/feature_transforms.py

```diff
@@ -0,0 +1,713 @@
+
+  placeholder = tf.placeholder(dtype=tf.string, shape=(None,),
+                               name='csv_input_placeholder')
+  tensors = tf.decode_csv(placeholder, record_defaults)
+  raw_features = dict(zip(csv_header, tensors))
+
+  transform_fn = make_preprocessing_fn(analysis_path, features, keep_target)
+  transformed_tensors = transform_fn(raw_features)
+
+  transformed_features = {}
+  for k, v in six.iteritems(transformed_tensors):
+    if isinstance(v, tf.Tensor) and v.get_shape().ndims == 1:
+      transformed_features[k] = tf.expand_dims(v, -1)
+    else:
+      transformed_features[k] = v
+
+  return input_fn_utils.InputFnOps(
+      transformed_features, None, {"csv_example": placeholder})
+
+
+def build_csv_transforming_training_input_fn(schema,
+                                             features,
+                                             analysis_output_dir,
+                                             raw_data_file_pattern,
+                                             training_batch_size,
+                                             num_epochs=None,
+                                             randomize_input=False,
+                                             min_after_dequeue=1,
+                                             reader_num_threads=1):
+  """Creates training input_fn that reads raw csv data and applies transforms.
+
+  Args:
+    schema: schema list
+    features: features dict
+    analysis_output_dir: output folder from analysis
+    raw_data_file_pattern: file path, or list of files
+    training_batch_size: An int specifying the batch size to use.
+    num_epochs: numer of epochs to read from the files. Use None to read forever.
+    randomize_input: If true, the input rows are read out of order. This
+        randomness is limited by the min_after_dequeue value.
+    min_after_dequeue: Minimum number elements in the reading queue after a
+        dequeue, used to ensure a level of mixing of elements. Only used if
+        randomize_input is True.
+    reader_num_threads: The number of threads enqueuing data.
+
+  Returns:
+    An input_fn suitable for training that reads raw csv training data and
+    applies transforms.
+
+  """
+
+  def raw_training_input_fn():
+    """Training input function that reads raw data and applies transforms."""
+
+    if isinstance(raw_data_file_pattern, six.string_types):
+      filepath_list = [raw_data_file_pattern]
+    else:
+      filepath_list = raw_data_file_pattern
+
+    files = []
+    for path in filepath_list:
+      files.extend(file_io.get_matching_files(path))
+
+    filename_queue = tf.train.string_input_producer(
+        files, num_epochs=num_epochs, shuffle=randomize_input)
+
+    csv_id, csv_lines = tf.TextLineReader().read_up_to(filename_queue, training_batch_size)
+
+    queue_capacity = (reader_num_threads + 3) * training_batch_size + min_after_dequeue
+    if randomize_input:
+      batch_csv_id, batch_csv_lines = tf.train.shuffle_batch(
```

Comment:
Correct, but we still need to pass a list to shuffle_batch and batch, and remove the returned list. So we would need to do

batch_csv_lines, = tf.train.shuffle_batch(tensors=[csv_lines], ...)

but then the ',' looks odd and needs a comment. So instead I'll just replace batch_csv_id with '_'